core/arch/interp.c

```diff
@@ -817,6 +817,18 @@ check_new_page_jmp(dcontext_t *dcontext, build_bb_t *bb, app_pc new_pc)
 }
 
 static inline void
+bb_process_single_step(dcontext_t *dcontext, build_bb_t *bb)
+{
+    LOG(THREAD, LOG_INTERP, 2, "interp: single step exception bb at "PFX"\n", bb->instr_start);
+    instrlist_append(bb->ilist, bb->instr);
+
+    /* Mark instruction as special exit. */
+    instr_branch_set_special_exit(bb->instr, true);
```

Comment:
Also, we want to delete it afterward: we don't want someone entering it without having hit an exception first.